### ISSUE TITLE
Project source code management (.gitignore file managment) and this tool

### ISSUE THREAD

**Issue by alankent** [SnowdogApps/magento2-frontools#11]
Great tool, still loving it. One issue I have however is we are moving towards projects only putting 'vendor' in '.gitignore' for the rest of Magento 2. (Its going to take a while to get there, but it is the target.) This tool however violates this rule. The problems I have are:

- Files I should not change do not go under 'vendor'.
- Files I should change (config files) are mixed in with files I should not change - so its not easy to include files I should change in my project GIT repo but not commit files I should not change.
- `npm install` puts LOTS of big files under the tool directory. That means you cannot just ignore the problem and commit the whole directory to your project's GIT repo.
- Currently this picks `tools` has its home directory - should it be called `frontools` to match the project name to avoid confusion? (So config files go in say `frontools/configs` in the GIT project tree.)

Suggestion (might be better ideas):

- Default recommendation to install this tool is its own directory/composer project. This is a tool what works on multiple projects, and it has files you should not change per project, so install it more like a tool.
- Have the tool to load configuration files relative to the directory `gulp` is run from with an environment variable `FRONTOOL_HOME` (or `~/.frontool`) or similar holding the directory to where the tool is, or a shell script I can put in my `PATH` variable so the `configs` directory can exist per M2 project independent to the rest of the tool.

Basically make it easy to install the tool somewhere shared, but save the per project configuration in the GIT repo for that project.

And if you get it going nicely, let me know so we can copy for our built in grunt support - I am thinking to remove grunt support to be a separate project from the M2 code base, just like this tool is. It has exactly the same problem at the moment.

**Comment by Igloczek** [comment]
Why not to add another directory inside `.gitignore`? 
Simply ignore `tools` (or as you said `frontools`) or `node_modules` and commit to main repo -> done, you can do whatever you want, without adding unnecessary complexity.

One frontools instance = one Magento 2 project. Most of projects requires some kind of tools customization, so there is no way to do this with single code base. Also, how to keep few tools versions at same time and be sure to have same version on every environment?

I don't want to make this hard to use, so idea to move tools outside projects directory or to require some system configuration is agains simplicity.

**Comment by alankent** [comment]
There are a few different things going on here, so I will try to break them down one by one.

1. One aspect of this is that I want some files under "tools" to be in my git project - for example, I want to commit "tools/configs/themes.json" in the project. I don't want that to be lost, I don't want to type it in per person working on a project. If I .gitignore the whole directory tree then I cannot save my configuration settings for your tool on a per project basis.

2. To further back up point 1, I think it makes sense having a build process to construct the final build of a site. Some frontend devs has said they want to use exactly the same CSS files during development in production. That is, they want to use Gulp output in production as well. That means the build process needs exactly the same configuration settings as local development. The build process is not always done by the developer - some sites have it so you do a commit to a Git repo, then that triggers a CICD run and build a new Docker image or similar if all tests pass. So even if you are not 100% convinced by point 1, a build process done by CICD needs the tools/configs/themes.json file committed into the GIT repo.

3. To further back up the build process approach triggered by a commit to GIT, you don't really want to reinstall Gulp every time you do a build. A "clean build" does a git clone into an empty directory, runs all the tests and compile commands, then if everything is green marks the final build as "good". So to do a good CICD solution I do not want to reinstall your tool each time if I can avoid it, because it is slow. I want to set up my CICD to be as fast as possible, so each commit I can run tests automatically. Having the tool installation inside the project source tree makes this hard to optimize. Its much easier if the tool is outside the project directory with only the configuration files needed by the tool inside the project directory.

But you are right in that there are still challenges - such as how to make sure the tool is the same version as the project. This is a problem. It is a matter of selecting between the lesser of evils. Do we (you!) recommend reinstalling npm, gulp, etc per push to the GIT repo (assuming a CICD run is triggered per git commit)? If not, then need to separate the tool from the tool configuration. This is where "good" tools put a version number in the configuration file and are backwards compatible as long as possible. If the config file is newer than the tool, the tool throws an error; if the tool is newer than the config file, the tool implements the semantics as per that older version number (throwing an exception if it cannot do so anymore).

Why are we trying to push things out of the root directory? To clean things up. Vendor is designed for 3rd party files - that is its sole purpose for existence. So if the tool was to be installed per project, then it makes more sense it going under 'vendor' - except that it *also* contains project settings.

Or do you think "project" build settings should be kept separate to "developer" settings. That is, developers fiddle a lot. An official "build" should not use the developer settings - a separate config file should be used for those settings as they are more stable.

Does this make sense?

**Comment by Igloczek** [comment]
1.  `/tools` is only working copy of composer package inside `/vendor` (we don't want to dig inside vendor folder few times per day, to run tools we need), so if you want to make some changes, you have to edit source code inside `vendor` (unfortunately tools cannot be symlinked).
2. Sure, I said this too :) Workflow with this tools looks like this:
  * fork repo
  * add as composer module to single project
  * make changes (inside `/vendor`) and commit them
  * use wherever you want
3. I just repeat my question "How to keep few tools versions at same time and be sure to have same version on every environment?" and add another point "How to keep proper versions of tools on CICD workers, if they works on many branches (and possibly many version of tools) or even projects (Docker)?" Performance of build process it's CICD pipeline problem. I'm not familiar with all of them, but in one we use is an options to cache specified folders between builds, so dependencies don't have to be downloaded each time. 

I'm not building enterprise grade solution, with LTS versions, professional support etc. 
I just care about one thing - ease of use. We (as front-devs) use this kind of tools a lot, so it should be as close to root project dir as possible.

We already have a loot to learn on M2, so adding another level of complexity, just to keep rules that are against this idea, is not way that I want to follow. 
PS. Almost everyday I see fear into eyes of M2 devs and want to help them :)

**Comment by alankent** [comment]
I am mulling through this one myself - we have the same problem with Grunt. What I am hearing is people want to use it for the official build process as well as local development, to guarantee what you develop/test is what you use in production. But the nodejs libraries are big, so you don't want node_packages/ in git - .gitignore to the rescue I guess.

It seems common to reinstall the nodejs packages per Jenkins build, relying on NodeJS caches to avoid downloading too much each time. That is, the build process is something like:

```
git clone ...            # Get fresh directory tree of code
composer install         # Get modules
npm install              # Get NodeJS modules
gulp styles              # Compile the styles
magento setup:static-content:deploy.    # Well, not quite, as gulp does much of this now. (TODO!)
magento di:compiler.     # Code generation
```

Some people talk about how this slows down the build process (e.g. can take 3 minutes to install node js packages per build). https://lincolnloop.com/blog/speeding-npm-installs/  talks about an optimization process to get this faster. But you are right in that all the JS + Jenkins discussion talks about running npm install each build.

But the primary goal was to see if fontools can be made to run like the rest of Magento (will). That was to try and simplify things for developers.

Another problem with the above is the composer install can override any local changes, hence the desire to keep per project files and shared files (read-only installed by gulp) separate. Otherwise the build process has to save away locally modified files somehow, and put them back again after running composer install.

More thinking required to get automated build processes working with Gulp nicely.

**Comment by Igloczek** [comment]
> More thinking required to get automated build processes working with Gulp nicely.

For sure, this is just beginning and there might be some changes in near future (when we in Snowdog tries to set up first CICD pipeline for M2 - soon :wink: ), but it looks like most of problems/things to do is  in CICD tool of choice configuration, not frontools itself.

**Comment by Igloczek** [comment]
Looks like we can close this issue.
